atproto/crypto/keys.go

```diff
@@ -0,0 +1,457 @@
+	case K256:
+		return k.privK256.Equal(other.privK256)
+	default:
+		panic("unexpected crypto KeyType")
+	}
+}
+
+func (k *PrivateKey) KeyType() KeyType {
+	return k.keyType
+}
+
+// Serializes the secret key material in to a raw binary format, which can be parsed by [ParsePrivateKeyBytes].
+//
+// The encoding format is curve-specific, and is generally "compact" for private keys. Both P-256 and K-256 private keys end up 32 bytes long. There is no ASN.1 or other enclosing structure to the binary encoding.
+func (k *PrivateKey) Bytes() ([]byte, error) {
+	switch k.keyType {
+	case P256:
+		skEcdh, err := k.privP256.ECDH()
+		if err != nil {
+			return nil, fmt.Errorf("unexpected failure to convert key type: %w", err)
+		}
+		return skEcdh.Bytes(), nil
+	case K256:
+		return k.privK256.Bytes(), nil
+	default:
+		return nil, fmt.Errorf("unexpected crypto KeyType")
+	}
+}
+
+// Outputs the PublicKey corresponding to this PrivateKey.
+func (k *PrivateKey) Public() PublicKey {
+	switch k.keyType {
+	case P256:
+		return PublicKey{
+			keyType: k.keyType,
+			pubP256: k.privP256.Public().(*ecdsa.PublicKey),
+		}
+	case K256:
+		return PublicKey{
+			keyType: k.keyType,
+			pubK256: k.privK256.PublicKey(),
+		}
+	default:
+		panic("unexpected crypto KeyType")
+	}
+}
+
+// First hashes the raw bytes, then signs the digest, returning a binary signature.
+//
+// SHA-256 is the hash algorithm used, as specified by atproto. Signing digests is the norm for ECDSA, and required by some backend implementations. This method does not "double hash", it simply has name which clarifies that hashing is happening.
+//
+// Calling code is responsible for any string encoding of signatures (eg, hex or base64). Both P-256 and K-256 signatures are 64 bytes long.
+//
+// NIST ECDSA signatures can have a "malleability" issue, meaning that there are multiple valid signatures for the same content with the same signing key. This method always returns a "low-S" signature, as required by atproto.
+func (k *PrivateKey) HashAndSign(content []byte) ([]byte, error) {
+	hash := sha256.Sum256(content)
+	switch k.keyType {
+	case P256:
+		r, s, err := ecdsa.Sign(rand.Reader, k.privP256, hash[:])
+		if err != nil {
+			return nil, fmt.Errorf("crypto error signing with P-256/secp256r1 private key: %w", err)
+		}
+		s = sigSToLowS_P256(s)
+		sig := make([]byte, 64)
+		r.FillBytes(sig[:32])
+		s.FillBytes(sig[32:])
+		return sig, nil
+	case K256:
+		return k.privK256.Sign(rand.Reader, hash[:], k256Options)
+	default:
+		return nil, fmt.Errorf("unexpected crypto KeyType")
+	}
+}
+
+// Checks if the two public keys are the same. Note that the naive == operator does not work for most equality checks.
+func (k *PublicKey) Equal(other *PublicKey) bool {
+	if k.keyType != other.keyType {
+		return false
+	}
+	switch k.keyType {
+	case P256:
+		return k.pubP256.Equal(other.pubP256)
+	case K256:
+		return k.pubK256.Equal(other.pubK256)
+	default:
+		panic("unexpected crypto KeyType")
```

Comment:
Again, do we need to panic here or can we return an error from this function?

I know it's not as "clean" from a equality operator standpoint but it may be more helpful.